File: dynamic-programming/leetcode64.py
#!/usr/bin/env python
# -*- coding:utf-8 -*-
"""
    @Header	leetcode64.py
    @Author	fosaken
    @Time	2020/12/19 16:47:00
    Abstract:
"""

from typing import List


class Solution:
    def minPathSum(self, grid: List[List[int]]) -> int:
        h = len(grid)
        w = len(grid[0]) if h > 0 else 0
        if not h or not w:
            return 0
        dp = [[0] * w for _ in range(h)]
        dp[0][0] = grid[0][0]
        for i in range(1, h):
            dp[i][0] = dp[i-1][0] + grid[i][0]
        for i in range(1, w):
            dp[0][i] = dp[0][i-1] + grid[0][i]
        for i in range(1, h):
            for j in range(1, w):
                dp[i][j] = min(dp[i-1][j], dp[i][j-1]) + grid[i][j]
        return dp[-1][-1]


def main():
    grid = [[1,4,8,6,2,2,1,7],[4,7,3,1,4,5,5,1],[8,8,2,1,1,8,0,1],[8,9,2,9,8,0,8,9],[5,7,5,7,1,8,5,5],[7,0,9,4,5,6,5,6],[4,9,9,7,9,1,9,0]]
    solution = Solution()
    result = solution.minPathSum(grid)
    print(result)


if __name__ == "__main__":
    main()

File: dynamic-programming/leetcode64.cpp
#include <iostream>
#include <vector>
#include <numeric>

using namespace std;
// int help(vector<vector<int>>& grid, int i, int j){
//     if(i == 0){
//         int re = 0;
//         for (int s = 0; s <= j; s++){
//             re += grid[0][s];
//         }
//         return re;
//     }
//     if (j == 0){
//         int re = 0;
//         for (int s = 0; s <= i; s++){
//             re += grid[s][0];
//         }
//         return re;
//     }

//     return min(help(grid, i - 1, j), help(grid, i, j - 1)) + grid[i][j];
// }
// int minPathSum(vector<vector<int>>& grid){
//     return help(grid, grid.size()-1, grid[0].size()-1);
// }


// int helper(vector<vector<int>>& grid, vector<vector<int>>& grid_sum, int i, int j){
//         if(i == 0){
//         int re = 0;
//         for (int s = 0; s <= j; s++){
//             re += grid[0][s];
//         }
//         return re;
//     }
//     if (j == 0){
//         int re = 0;
//         for (int s = 0; s <= i; s++){
//             re += grid[s][0];
//         }
//         return re;
//     }
//     if (grid_sum[i-1][j] == -1){
//         grid_sum[i - 1][j] = helper(grid, grid_sum, i - 1, j);
//     }
//     if (grid_sum[i][j-1] == -1){
//         grid_sum[i][j-1] = helper(grid, grid_sum, i, j-1);
//     }
//     return min(grid_sum[i][j-1], grid_sum[i - 1][j]) + grid[i][j];
// }

// int helper(vector<vector<int>>& grid, vector<vector<int>>& grid_sum, int i, int j){
//     if(i == 0){
//         int re = 0;
//         for (int s = 0; s <= j; s++){
//             re += grid[0][s];
//         }
//         return re;
//     }
//     if (j == 0){
//         int re = 0;
//         for (int s = 0; s <= i; s++){
//             re += grid[s][0];
//         }
//         return re;
//     }
//     // if (grid_sum[i-1][j] == -1){
//     //     grid_sum[i - 1][j] = helper(grid, grid_sum, i - 1, j);
//     // }
//     // if (grid_sum[i][j-1] == -1){
//     //     grid_sum[i][j-1] = helper(grid, grid_sum, i, j-1);
//     // }
//     if (grid_sum[i][j] == -1){
//         grid_sum[i][j] = min(helper(grid, grid_sum, i, j - 1), helper(grid, grid_sum, i - 1, j)) + grid[i][j];
//     }
//     return  grid_sum[i][j];
// }


int minPathSum(vector<vector<int>>& grid){
    vector<vector<int>> grid_sum(grid.size(), vector<int>(grid[0].size(), -1));
    grid_sum[0][0] = grid[0][0];
    for (int m = 1; m < grid.size(); m++){
        grid_sum[m][0] = grid_sum[m - 1][0] + grid[m][0];
    }
    for (int m = 1; m < grid[0].size(); m++){
        grid_sum[0][m] = grid_sum[0][m - 1] + grid[0][m];
    }
    for (int m = 1; m < grid.size(); m++)
    {
        for (int n = 1; n < grid[0].size(); n++)
        {
            grid_sum[m][n] = min(grid_sum[m][n - 1], grid_sum[m - 1][n]) + grid[m][n];
        }
    }
    return  grid_sum[grid.size()-1][grid[0].size()-1];
}

int main(){
    vector<vector<int>> grid = {{1,3,1},{1,5,1},{4,2,1}};
    int re = minPathSum(grid);
    cout << re;
    system("pause");
    return 0;
}